### PR TITLE
Add support for Placement directives in applications and add a new Machine resource

### PR DESCRIPTION
**Pull request by jadonn**
## Description

This PR implements two parts of the Juju's functionality. The first part is defining a placement directive in Charm applications to instruct Juju where to deploy Charm units (this is the equivalent of the `--to` Juju CLI argument). The second part adds a new Machine resource that allows you to use Juju to create machines and manage them through Terraform (this is meant to do the same thing as the `add-machine` command in Juju).

I wrote this extension to the Terraform Juju provider because I wanted to use the Terraform Juju provider to deploy Charmed OpenStack as outlined in the [official upstream Charmed OpenStack deployment guide](https://docs.openstack.org/project-deploy-guide/charm-deployment-guide/latest/install-openstack.html). I could not do that without placement directives because Juju would create a new machine for every application. My aim is to deploy OpenStack on as few machines as possible for efficiency. I added the placement directive to the Application resource schema and Application client alongside the existing arguments that are passed to the Juju client.

However, I realized that the placement directive requires that you know the machine IDs ahead of time. I was not going to know the machine IDs ahead of time because I wanted to dynamically create and destroy the machines I use to run OpenStack. I chose to implement a basic Machine resource in the Terraform Juju provider so that I can create the manages and manage them in Terraform alongside the applications. With a Machine resource, I can dynamically get the Juju machine ID and pass that into the Application resource definitions.

The code I added to the Terraform Juju provider works. I can use Terraform to deploy Charmed OpenStack on four machines with storage, compute, and control plane services on the same machines (that is, a hyperconverged deployment). I have [a Terraform plan in a rough draft state](https://github.com/jadonn/terraform-juju-openstack-testing/blob/main/main.tf) in a personal GitHub repo that will deploy everything for OpenStack using MAAS.

The overall goal for this work is to orchestrate automating the deployment of OpenStack so I can set up frequent automated testing of OpenStack. I really like the Terraform Juju provider for this project because I do not have to write a great deal of custom code that would only ever be used for this one project. Using the Terraform Juju provider significantly reduces my overhead and simplifies the configuration of my deployment considerably.

## Type of change

Please mark if proceeds.

- [X] New resource (a new resource in the schema)
- [X] Changed resource (changes in an existing resource)
- [ ] Logic changes in resources (the API interaction with Juju has been changed)
- [X] Test changes (one or several tests have been changed)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update
- [ ] Other (describe)

## Environment

- Juju controller version: 2.9
- Terraform version: 1.3

## QA steps

You should be able to test this PR with any Juju cloud. During development, I used a LXD cloud running on my local machine to test provisioning machines and using the placement directive. You should be able to run the following simple plan to verify the core functionality of creating a machine and assigning units with the placement directive

```tf
provider juju {}

resource "juju_model" "machine_test" {
  name = "machine-test"
}

resource "juju_machine" "this" {
  depends_on = [juju_model.machine_test]
  model = juju_model.machine_test.name
  series = "focal"
  name = "testMachine"
}

resource "juju_machine" "that" {
  model = juju_model.machine_test.name
  series = "focal"
  name = "thatMachine"
}

resource "juju_application" "hello" {
  model = juju_model.machine_test.name
  charm {
    name = "hello-juju"
  }
  placement = join(",", [split(":", juju_machine.this.id)[1], split(":", juju_machine.that.id)[1]])
  units = 2
}
```

This plan should create two machines using Ubuntu Focal and deploy one unit of the `hello-juju` application to each machine. After the plan runs, you should see the correct machines and placement when you use `juju status` to see the model.

# Additional notes

Let me know what changes need to be made or if you would like to take a different approach to representing machines or placement directives. I am happy to make changes to follow your overall plans or goals better.

### CODE REVIEW

File: internal/provider/resource_application.go [hunk 1]
```diff
@@ -127,6 +127,11 @@ func resourceApplication() *schema.Resource {
 					},
 				},
 			},
+			"placement": {
+				Description: "Specify the target location for the application's units",
+				Type: schema.TypeString,
+				Optional: true,
+				Default: "",
```
Comment (gboutry) [inline internal/provider/resource_application.go:134]:
```suggestion
			"placement": {
				Description: "Specify the target location for the application's units",
				Type: schema.TypeString,
				Optional: true,
				Default: "",
			},
```

Merging main into this branch broke the code

Comment (jadonn) [reply internal/provider/resource_application.go:134]: I'll take a look at what happened and submit fixes for what is preventing the build. This is what is happening when you try to build with the merged main code:
```
$ go build -o terraform-provider-juju
go: downloading github.com/rs/zerolog v1.29.0
go: downloading github.com/hashicorp/terraform-plugin-sdk/v2 v2.25.0
go: downloading github.com/hashicorp/go-hclog v1.4.0
go: downloading github.com/hashicorp/go-plugin v1.4.8
go: downloading github.com/hashicorp/terraform-plugin-go v0.14.3
go: downloading github.com/hashicorp/terraform-plugin-log v0.8.0
go: downloading golang.org/x/crypto v0.6.0
go: downloading golang.org/x/net v0.6.0
go: downloading google.golang.org/grpc v1.51.0
go: downloading golang.org/x/text v0.7.0
go: downloading golang.org/x/sys v0.5.0
go: downloading github.com/juju/worker/v3 v3.1.0
go: downloading github.com/hashicorp/terraform-registry-address v0.1.0
go: downloading github.com/hashicorp/hcl/v2 v2.16.1
go: downloading golang.org/x/term v0.5.0
# github.com/juju/terraform-provider-juju/internal/provider
internal/provider/resource_application.go:141:3: syntax error: unexpected newline in composite literal; possibly missing comma or }
internal/provider/resource_application.go:144:6: syntax error: unexpected resourceApplicationCreate, expecting (
internal/provider/resource_application.go:168:2: syntax error: non-declaration statement outside function body
internal/provider/resource_application.go:184:2: syntax error: non-declaration statement outside function body
```

File: internal/juju/applications.go [hunk 2]
```diff
@@ -614,6 +647,7 @@ func (c applicationsClient) ReadApplication(input *ReadApplicationInput) (*ReadA
 		Config:      conf,
 		Constraints: appConstraints,
 		Principal:   appInfo.Principal,
+    Placement: placement,
```
Comment (jnsgruk) [inline internal/juju/applications.go:650]: uber nit: probably needs a pass with `gofmt` :)

**Comment by jnsgruk** [comment]
Thank you for tackling this @jadonn!

I do wonder if this should be split into 2 PRs though:

- addition of the machines api
- ability to set placement

My sense is that the machines API work will get us slightly closer to implementing `juju add-machine` type workflows.

WDYT @juanmanuel-tirado?

**Comment by juanmanuel-tirado** [comment]
It is true that the placement directive is easy to add in the application API and does not really need to have the machine resource to be done. I do agree. It would probably be better organized if split into two PRs.

**Comment by jadonn** [comment]
Thanks for the feedback! Splitting it into two PRs works for me from a code standpoint. I implemented the two together because for my use case of deploying OpenStack it felt like also deploying the machines through Terraform would be an effective way of obtaining the information I needed from Juju to then use the Placement directive in applications.

I will work on splitting the PR, run go fmt, and fix the breaks from merging the commits I was missing from the main branch.

File: internal/juju/applications.go (hunk 2 above)
Comment (jadonn) [reply internal/juju/applications.go]: I ran gofmt and made the changes it recommended. Thanks for the feedback!

**Comment by jadonn** [comment]
I reworked this PR to add just the placement directive to the Application Resource. I believe I have separated out the machine resource code, and this should be ready for review. I will get the other PR submitted for the machine resource done this week.

**Review by juanmanuel-tirado — CHANGES_REQUESTED** [review]
`go generate` has to be run

I'm OK with the PR. Could you add an example of the placement in the `examples/resources/juju_application/resource.tf`?

**Comment by jadonn** [comment]
Thanks! I added an example of placement in the file you indicated, and I ran `go generate` as you asked. As far as I can tell with the last commit the submission should pass the test that was failing.

**Review by juanmanuel-tirado — APPROVED** [review]
LG2M

Thanks

**Comment by jnsgruk** [comment]
Nice! Thank you @jadonn!

**Comment by jadonn** [comment]
Thank you all for your support and for accepting my PR! I really appreciate the work you are doing and the opportunity to contribute back.

I have really enjoyed using the Terraform Juju provider so far, and I look forward to building on top of it

**Pull request merged**